pkg/squash/vfs_fs.go

```diff
@@ -0,0 +1,210 @@
+package squash
+
+import (
+	"archive/tar"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"sort"
+	"strings"
+	"sync"
+	"syscall"
+)
+
+var (
+	_ fs.FS          = (*fsfile)(nil)
+	_ fs.File        = (*fsfileReader)(nil)
+	_ fs.ReadDirFile = (*fsfileReader)(nil)
+)
+
+var (
+	ErrIsDir   = syscall.EISDIR
+	ErrMissing = errors.New("not in layers")
+)
+
+// Open implements io/fs.FS
+func (f *fsfile) Open(name string) (_ fs.File, err error) {
+	defer func() {
+		if err != nil {
+			err = &fs.PathError{
+				Op:   "open",
+				Path: name,
+				Err:  err,
+			}
+		}
+	}()
+	if !fs.ValidPath(name) {
+		return nil, fs.ErrInvalid
+	}
+	lnk, err := fsGet(f, name, false, false)
+	if err != nil {
+		return nil, err
+	}
+	tgt, err := lnk.Get(".", false, true)
+	if err != nil {
+		return nil, err
+	}
+	return &fsfileReader{ //nolint:exhaustivestruct
+		origName: name,
+		tgt:      tgt,
+		lnk:      lnk,
+	}, nil
+}
+
+type fsfileReader struct {
+	origName string
+	tgt      *fsfile
+	lnk      *fsfile
+
+	mu      sync.Mutex
+	pos     int
+	closed  bool
+	dirents []fs.DirEntry
+}
+
+// Open implements io/fs.File
+func (f *fsfileReader) Stat() (fs.FileInfo, error) {
+	if f.tgt.header == nil {
+		return nil, &fs.PathError{
+			Op:   "stat",
+			Path: f.origName,
+			Err:  ErrMissing,
+		}
+	}
+	hdr := *f.tgt.header // shallow copy
+	hdr.Name = f.lnk.header.Name
+	return hdr.FileInfo(), nil
+}
+
+// Open implements io/fs.File
+func (f *fsfileReader) Read(buf []byte) (_ int, err error) {
+	defer func() {
+		if err != nil && !errors.Is(err, io.EOF) {
+			err = &fs.PathError{
+				Op:   "read",
+				Path: f.origName,
+				Err:  err,
+			}
+		}
+	}()
+	if f.tgt.header == nil || f.tgt.header.Typeflag == tar.TypeDir {
+		return 0, ErrIsDir
+	}
+	if int64(len(f.tgt.body)) < f.tgt.header.Size {
+		return 0, ErrMissing
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.closed {
+		return 0, fs.ErrClosed
+	}
+	if int64(len(f.tgt.body)) < f.tgt.header.Size {
+		return 0, ErrMissing
+	}
+	if f.pos == len(f.tgt.body) {
+		return 0, io.EOF
+	}
+	n := copy(buf, f.tgt.body[f.pos:])
+	f.pos += n
+	return n, nil
+}
+
+// Open implements io/fs.File
```

Comment:
Yet another `Open` mis-comment.

Also... `Close` here really _doesn't_ implement `io/fs.File`. It is part of the implementation of `io/fs.File`. Shouldn't this really say that `fsFileReader` implements `io/fs.File`?